C/C++ Data Sizes

In msvc or gcc on windows, data sizes are different from gcc on regular linux platforms.
Particularly, `long int` is 4 bytes, and `long long int` is 8 bytes.

```console
> make
...
g++ -DDEBUGGING_HACKS -Wall -Werror -std=c++11 -I.  -c -o memory.o memory.cc
memory.cc: In member function 'void memory::dump(uint64_t, uint64_t)':
memory.cc:198:26: error: format '%lx' expects argument of type 'long unsigned int', but argument 3 has type 'uint64_t' {aka 'long long unsigned int'} [-Werror=format=]
  198 |   fprintf(ddtout, " %8.8lx:", addr&~0x0f);
      |                     ~~~~~^    ~~~~~~~~~~
      |                          |        |
      |                          |        uint64_t {aka long long unsigned int}
      |                          long unsigned int
...
```

As stated [here](https://software.intel.com/content/www/us/en/develop/articles/size-of-long-integer-type-on-different-architecture-and-os.html) for windows, linux, and mac:
> **Suggestion**:   If it is important to you for integer types to have the same size on all Intel platforms, then consider replacing "long" by either "int" or "long long". The size of the "int" integer type is 4 bytes and the size of the "long long" integer type is 8 bytes for all the above combinations of operating system and architecture.

And if i correctly understand the table from [this page](https://en.cppreference.com/w/cpp/io/c/fprintf), you can replace the modifier `l` - `long` by `ll` - `long long`, so the format string would be `" %8.8llx:"`, and this will make gcc on windows\linux-x86 happy again.

Yeah.  My mistake.  This app should have been done using cout & 
insertion operator.

When I have time I'll have to rewrite it.

There are several other simulators out there if you just one that you 
can use.  I am not sure if there are any interactive ones that are as 
simplistic as rvddt though.   Definitely not when I wrote it.


-John

On 10/16/20 1:07 PM, myCrack wrote:
> In msvc or gcc on windows, data sizes are different from gcc on regular 
> linux platforms. Particularly, |long int| is 4 bytes, and |long long 
> int| is 8 bytes.
> 
>> make
> ...
> g++ -DDEBUGGING_HACKS -Wall -Werror -std=c++11 -I. -c -o memory.o memory.cc
> memory.cc: In member function 'void memory::dump(uint64_t, uint64_t)':
> memory.cc:198:26: error: format '%lx' expects argument of type 'long 
> unsigned int', but argument 3 has type 'uint64_t' {aka 'long long 
> unsigned int'} [-Werror=format=]
> 198 | fprintf(ddtout, " %8.8lx:", addr&~0x0f);
> | ~~~~~^ ~~~~~~~~~~
> | | |
> | | uint64_t {aka long long unsigned int}
> | long unsigned int
> ...
> 
> As stated here 
> <https://software.intel.com/content/www/us/en/develop/articles/size-of-long-integer-type-on-different-architecture-and-os.html> 
> for windows, linux, and mac:
> 
>     *Suggestion*: If it is important to you for integer types to have
>     the same size on all Intel platforms, then consider replacing "long"
>     by either "int" or "long long". The size of the "int" integer type
>     is 4 bytes and the size of the "long long" integer type is 8 bytes
>     for all the above combinations of operating system and architecture.
> 
> And if i correctly understand the table from this page 
> <https://en.cppreference.com/w/cpp/io/c/fprintf>, you can replace the 
> modifier |l| - |long| by |ll| - |long long|, so the format string would 
> be |" %8.8llx:"|, and this will make windows-gcc happy again.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/johnwinans/rvddt/issues/4>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABQ6H766S3Z2XDHB6TEFAGLSLCDUNANCNFSM4STUWNPQ>.
>

--- a/src/memory.cc
+++ b/src/memory.cc
@@ -195,7 +195,7 @@ void memory::dump(uint64_t addr, uint64_t length)
 	if (startingoffset != 0)
 	{
 		// pre-padd the dump so it looks purdy
-		fprintf(ddtout, " %8.8lx:", addr&~0x0f);
+		fprintf(ddtout, " %8.8llx:", addr&~0x0f);
 		int i;
 		for (i=0; i<startingoffset; ++i)
 			fprintf(ddtout, "   ");

